File: competitive-programming/src/edu/algos/li04_slidingwindow/easy/L643_MaximumAverageSubarrayI.java
package edu.algos.li04_slidingwindow.easy;

public class L643_MaximumAverageSubarrayI {

	public static void main(String[] args) {
		L643_MaximumAverageSubarrayI obj = new L643_MaximumAverageSubarrayI();
		
		int[] nums = new int[] {1,12,-5,-6,50,3};
		System.out.println(obj.findMaxAverage(nums, 4));
	}
	
	/*
	 * Problem: https://leetcode.com/problems/maximum-average-subarray-i/
	 * Idea: Track the fixed size sliding window of k. Cal sum from it and avg. Track global min avg.
	 * */
	public double findMaxAverage(int[] nums, int k) {
        double sum = 0, avg = Integer.MIN_VALUE;
        for(int i=0; i<nums.length; i++){
            if(i<k){
                sum += nums[i];
                continue;
            }
            
            avg = Math.max(avg, sum/k);
            sum = sum - nums[i-k]+nums[i];
        }
        avg = Math.max(avg, sum/k);
        
        return avg;
    }

}

File: dsa/patterns/04_slidingwindow/python/l643_maximum_average_subarray_i.py
# L643 - Maximum Average Subarray I
#
# Problem: https://leetcode.com/problems/maximum-average-subarray-i/
#
# Idea:
# - Use a fixed-size sliding window of length k
# - Track the sum of the window and update the maximum average
#
# Time: O(n)
# Space: O(1)


def find_max_average(nums, k):
    window_sum = sum(nums[:k])
    max_sum = window_sum

    for i in range(k, len(nums)):
        window_sum += nums[i] - nums[i - k]
        max_sum = max(max_sum, window_sum)

    return max_sum / k


# -------------------- TEST --------------------
if __name__ == "__main__":
    nums = [1, 12, -5, -6, 50, 3]
    k = 4
    print(find_max_average(nums, k))  # Output: 12.75
